README.md
# mesg-pusher [![mesg-pusher](https://img.shields.io/badge/version-2.0.3-blue.svg)](https://github.com/Roms1383/mesg-pusher.git) [![mesg-core](https://img.shields.io/badge/mesg--core-0.9-blueviolet.svg)](https://github.com/mesg-foundation/engine.git) [![Build Status](https://travis-ci.com/Roms1383/mesg-pusher.svg?branch=master)](https://travis-ci.com/Roms1383/mesg-pusher) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

[MESG](https://docs.mesg.com) Service for [Pusher](https://pusher.com/docs/server_api_guide)

[![MESG](./logo-mesg.svg)](https://mesg.com/) | [![Pusher](./logo-pusher.svg)](https://pusher.com/)
--------------------------------------------- | ---------------------------------------------

# Contents

- [Installation](#Installation)
- [Features](#Features)
- [Definitions](#Definitions)
  
  - [Tasks](#Tasks)
    - [channel](#channel)
    - [channels](#channels)
    - [trigger](#trigger)
    - [triggerBatch](#triggerbatch)

# Installation

## MESG Core

This service requires [MESG Core](https://github.com/mesg-foundation/core) to be installed first.

You can install MESG Core by running the following command or [follow the installation guide](https://docs.mesg.com/guide/start-here/installation.html).

```bash
bash <(curl -fsSL https://mesg.com/install)
```

## Service

```bash
mesg-core service deploy https://github.com/Roms1383/mesg-pusher
```

# Features

Currently able to use `Pusher` [channels](https://pusher.com/docs/server_api_guide/interact_rest_api#application-channels), [channel](https://pusher.com/docs/server_api_guide/interact_rest_api#channel-information), [trigger](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events) and [triggerBatch](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-batches-of-events) features from `MESG`.

## Use case

`MESG Applications` on different servers communicating together over `Pusher` notifications.

### Example

1.  on `MESG Application` on a server `A` : use this `MESG Service` to send notifications over `Pusher`.

2.  on `MESG Application` on a server `B` : listen to the notifications emitted by `MESG Application` on a server `A` over `Pusher`.

## How to listen in a `MESG Application`

Install `MESG` and `Pusher` dependencies :

```shell
yarn add mesg-js pusher-js
```

Create a socket connection to listen and react to `Pusher` notifications :

```js
// in a MESG Application
const MESG = require('mesg-js').application()
const Pusher = require('pusher-js')
const pusher = new Pusher('PUSHER_APP_KEY', { cluster: 'PUSHER_CLUSTER', forceTLS: true }) // replace with your credentials
const CHANNEL = 'some-channel'
const EVENT = 'some-event'
const channel = pusher.subscribe(CHANNEL)
const CREDENTIALS = { appId: 'PUSHER_APP_ID', key: 'PUSHER_APP_KEY', secret: 'PUSHER_APP_SECRET' } // replace with your credentials
// on Pusher notification received
channel.bind(EVENT, data => {
  // example : launch MESG service trigger task
  MESG.api.ExecuteTask({
    serviceID: 'com.mesg.pusher',
    taskKey: 'trigger',
    inputData: JSON.stringify({ ...CREDENTIALS, channel: CHANNEL, event: EVENT, ...data })
  }, (err, reply) => {
    // handle response if needed
  })
})
```

## Environment variables

Instead of providing Pusher credentials or options on each request
e.g.
```js
...
const CREDENTIALS = { appId: 'PUSHER_APP_ID', key: 'PUSHER_APP_KEY', secret: 'PUSHER_APP_SECRET' } // replace with your credentials
MESG.api.ExecuteTask({
  serviceID: 'com.mesg.pusher',
  taskKey: 'trigger',
  inputData: JSON.stringify({ ...CREDENTIALS, channel: CHANNEL, event: EVENT, ...data })
}, (err, reply) => {
  // ...
})
...
```
These can be defined when deploying the service as environment variables :
```sh
mesg-core service deploy . --env PUSHER_APP_ID=YOUR_PUSHER_APP_ID --env PUSHER_APP_SECRET=YOUR_PUSHER_APP_SECRET --env PUSHER_APP_KEY=YOUR_PUSHER_APP_KEY
```

As usual, user inputs always takes precedence on defined environment variables.

| **Name** | **Enviroment Variable** |
| --- | --- |
| **appId** | `PUSHER_APP_ID` |
| **cluster** | `PUSHER_CLUSTER` |
| **useTLS** | `PUSHER_useTLS` |
| **host** | `PUSHER_HOST` |
| **keepAlive** | `PUSHER_KEEP_ALIVE` |
| **key** | `PUSHER_APP_KEY` |
| **port** | `PUSHER_PORT` |
| **proxy** | `PUSHER_PROXY` |
| **secret** | `PUSHER_APP_SECRET` |
| **timeout** | `PUSHER_TIMEOUT` |

## Integration tests

At the moment the integration tests are not yet supported correctly when run on `Travis`, but they can still be launched locally.

In order to launch them locally you will have to pass the environment variables along with your command, like so :

*  via command line arguments :
    ```bash
    mesg-core service stop com.mesg.pusher && mesg-core service deploy --env PUSHER_APP_ID=... --env    PUSHER_APP_KEY=... --env PUSHER_APP_SECRET=... --env PUSHER_CLUSTER=... && mesg-core service start com.mesg.pusher && yarn test
    ```
*  via `.env` file :
   *  file
      ```
      PUSHER_APP_ID=...
      PUSHER_APP_KEY=...
      PUSHER_APP_SECRET=...
      PUSHER_CLUSTER=...
      ```
   *  command
      ```bash
      mesg-core service stop com.mesg.pusher && mesg-core service deploy $(while read line; do echo "--env $line"; done < .env) && mesg-core service start com.mesg.pusher && yarn test
      ```
      or simply
      ```bash
      yarn mesg:test
      ```

Hopefully in the future they will :pray:

***

# Definitions


# Tasks

## channel

Task key: `channel`

fetch one or some attributes for a given channel

### Inputs

| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **appId** | `appId` | `String` | **`optional`** Pusher application ID |
| **cluster** | `cluster` | `String` | **`optional`** if `host` is present, it will override the `cluster` option |
| **useTLS** | `useTLS` | `Boolean` | **`optional`** whether to encrypt notification, defaults to false |
| **host** | `host` | `String` | **`optional`** whether to use a different host, defaults to api.pusherapp.com |
| **keepAlive** | `keepAlive` | `Boolean` | **`optional`** enables keep-alive, defaults to false |
| **key** | `key` | `String` | **`optional`** Pusher application key |
| **params** | `params` | `Object` | **`optional`** additional parameters to be sent as query string parameters (see [HTTP API Reference](https://pusher.com/docs/rest_api)) |
| **port** | `port` | `Number` | **`optional`** whether to use a different port, defaults to 80 for unuseTLS and 443 for useTLS |
| **proxy** | `proxy` | `String` | **`optional`** URL to proxy the requests through |
| **secret** | `secret` | `String` | **`optional`** Pusher application secret key |
| **timeout** | `timeout` | `Number` | **`optional`** timeout for all requests in milliseconds |

### Outputs

#### error

Output key: `error`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **error** | `error` | `String` | error |

#### success

Output key: `success`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **occupied** | `occupied` | `Boolean` | whether the channel currently has active subscriptions |
| **subscription_count** | `subscription_count` | `Number` | **`optional`** number of connections currently subscribed to this channel (not available by default, has to be enabled in dashboard) |
| **user_count** | `user_count` | `Number` | **`optional`** number of distinct users currently subscribed to this channel (a single user may be subscribed many times, but will only count as one) |


## channels

Task key: `channels`

get the list of the channel within an application that have active subscriptions (also referred to as being **occupied**)

### Inputs

| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **appId** | `appId` | `String` | **`optional`** Pusher application ID |
| **cluster** | `cluster` | `String` | **`optional`** if `host` is present, it will override the `cluster` option |
| **useTLS** | `useTLS` | `Boolean` | **`optional`** whether to encrypt notification, defaults to false |
| **host** | `host` | `String` | **`optional`** whether to use a different host, defaults to api.pusherapp.com |
| **keepAlive** | `keepAlive` | `Boolean` | **`optional`** enables keep-alive, defaults to false |
| **key** | `key` | `String` | **`optional`** Pusher application key |
| **params** | `params` | `Object` | **`optional`** additional parameters to be sent as query string parameters (see [HTTP API Reference](https://pusher.com/docs/rest_api)) |
| **port** | `port` | `Number` | **`optional`** whether to use a different port, defaults to 80 for unuseTLS and 443 for useTLS |
| **proxy** | `proxy` | `String` | **`optional`** URL to proxy the requests through |
| **secret** | `secret` | `String` | **`optional`** Pusher application secret key |
| **timeout** | `timeout` | `Number` | **`optional`** timeout for all requests in milliseconds |

### Outputs

#### error

Output key: `error`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **error** | `error` | `String` | error |

#### success

Output key: `success`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **channels** | `channels` | `Object` | array of channel names |


## trigger

Task key: `trigger`

triggers an event on one or more channels

### Inputs

| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **appId** | `appId` | `String` | **`optional`** Pusher application ID |
| **channel** | `channel` | `String` | **`optional`** channel name if publishing to a single channel (can be used instead of channels) |
| **channels** | `channels` | `Object` | **`optional`** array of one or more channel names - limited to 100 channels |
| **cluster** | `cluster` | `String` | **`optional`** if `host` is present, it will override the `cluster` option |
| **data** | `data` | `Object` | event data (maximum 10Kb) |
| **useTLS** | `useTLS` | `Boolean` | **`optional`** whether to encrypt notification, defaults to false |
| **host** | `host` | `String` | **`optional`** whether to use a different host, defaults to api.pusherapp.com |
| **keepAlive** | `keepAlive` | `Boolean` | **`optional`** enables keep-alive, defaults to false |
| **key** | `key` | `String` | **`optional`** Pusher application key |
| **name** | `name` | `String` | event name |
| **port** | `port` | `Number` | **`optional`** whether to use a different port, defaults to 80 for unuseTLS and 443 for useTLS |
| **proxy** | `proxy` | `String` | **`optional`** URL to proxy the requests through |
| **secret** | `secret` | `String` | **`optional`** Pusher application secret key |
| **socket_id** | `socket_id` | `Object` | **`optional`** excludes the event from being sent to a specific connection |
| **timeout** | `timeout` | `Number` | **`optional`** timeout for all requests in milliseconds |

### Outputs

#### error

Output key: `error`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **error** | `error` | `String` | error |

#### success

Output key: `success`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **message** | `message` | `String` | a dummy 'sent' message |


## triggerBatch

Task key: `triggerBatch`

triggers multiple events in a single call (up to 10 per call on the multi-tenant clusters)

### Inputs

| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **appId** | `appId` | `String` | **`optional`** Pusher application ID |
| **batch** | `batch` | `Object` | array of events (maximum 10) |
| **cluster** | `cluster` | `String` | **`optional`** if `host` is present, it will override the `cluster` option |
| **useTLS** | `useTLS` | `Boolean` | **`optional`** whether to encrypt notification, defaults to false |
| **host** | `host` | `String` | **`optional`** whether to use a different host, defaults to api.pusherapp.com |
| **keepAlive** | `keepAlive` | `Boolean` | **`optional`** enables keep-alive, defaults to false |
| **key** | `key` | `String` | **`optional`** Pusher application key |
| **port** | `port` | `Number` | **`optional`** whether to use a different port, defaults to 80 for unuseTLS and 443 for useTLS |
| **proxy** | `proxy` | `String` | **`optional`** URL to proxy the requests through |
| **secret** | `secret` | `String` | **`optional`** Pusher application secret key |
| **timeout** | `timeout` | `Number` | **`optional`** timeout for all requests in milliseconds |

### Outputs

#### error

Output key: `error`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **error** | `error` | `String` | error |

#### success

Output key: `success`



| **Name** | **Key** | **Type** | **Description** |
| --- | --- | --- | --- |
| **message** | `message` | `String` | a dummy 'sent' message |




CHANGELOG.md
## [2.0.3](https://github.com/Roms1383/mesg-pusher/compare/v2.0.2...v2.0.3) (2019-05-30)

## [2.0.2](https://github.com/Roms1383/mesg-pusher/compare/v2.0.1...v2.0.2) (2019-05-30)

## [2.0.1](https://github.com/Roms1383/mesg-pusher/compare/v2.0.0...v2.0.1) (2019-05-30)

# [2.0.0](https://github.com/Roms1383/mesg-pusher/compare/v1.1.0...v2.0.0) (2019-05-29)

# [1.1.0](https://github.com/Roms1383/mesg-pusher/compare/v1.0.14...v1.1.0) (2019-05-28)

## [1.0.14](https://github.com/Roms1383/mesg-pusher/compare/v1.0.13...v1.0.14) (2018-09-29)

## [1.0.13](https://github.com/Roms1383/mesg-pusher/compare/v1.0.12...v1.0.13) (2018-09-29)

## [1.0.12](https://github.com/Roms1383/mesg-pusher/compare/v1.0.11...v1.0.12) (2018-09-29)

## [1.0.11](https://github.com/Roms1383/mesg-pusher/compare/v1.0.10...v1.0.11) (2018-09-29)

## [1.0.10](https://github.com/Roms1383/mesg-pusher/compare/v1.0.9...v1.0.10) (2018-09-29)

## [1.0.9](https://github.com/Roms1383/mesg-pusher/compare/v1.0.8...v1.0.9) (2018-09-29)

## [1.0.8](https://github.com/Roms1383/mesg-pusher/compare/v1.0.7...v1.0.8) (2018-09-29)

## [1.0.7](https://github.com/Roms1383/mesg-pusher/compare/v1.0.6...v1.0.7) (2018-09-29)

## [1.0.6](https://github.com/Roms1383/mesg-pusher/compare/v1.0.5...v1.0.6) (2018-09-29)

## [1.0.5](https://github.com/Roms1383/mesg-pusher/compare/v1.0.4...v1.0.5) (2018-09-29)

## [1.0.4](https://github.com/Roms1383/mesg-pusher/compare/v1.0.3...v1.0.4) (2018-09-29)

## [1.0.3](https://github.com/Roms1383/mesg-pusher/compare/v1.0.2...v1.0.3) (2018-09-29)

## [1.0.2](https://github.com/Roms1383/mesg-pusher/compare/v1.0.1...v1.0.2) (2018-09-29)

## [1.0.1](https://github.com/Roms1383/mesg-pusher/compare/v1.0.0...v1.0.1) (2018-09-29)

# 1.0.0 (2018-09-29)


options.js
const yn = require('yn')
module.exports = ({
  appId = process.env.PUSHER_APP_ID,
  key = process.env.PUSHER_APP_KEY,
  secret = process.env.PUSHER_APP_SECRET,
  useTLS = yn(process.env.PUSHER_USE_TLS, { default: false }),
  cluster = process.env.PUSHER_CLUSTER || undefined,
  host = process.env.PUSHER_HOST || undefined,
  port = process.env.PUSHER_PORT || undefined,
  proxy = process.env.PUSHER_PROXY || undefined,
  timeout = process.env.PUSHER_TIMEOUT || undefined,
  keepAlive = yn(process.env.PUSHER_KEEP_ALIVE, { default: false })
}) => {
  const options = { appId, key, secret, useTLS, keepAlive }
  if (!appId) throw new Error(`options must contain application ID`)
  if (!key) throw new Error(`options must contain application key`)
  if (!secret) throw new Error(`options must contain application secret key`)
  if (!cluster && !host) throw new Error(`options must contain either cluster or host`)
  if (cluster) options.cluster = cluster
  if (host) options.host = host
  if (!port) options.port = useTLS ? 443 : 80
  if (proxy) options.proxy = proxy
  if (timeout) options.timeout = timeout
  return options
}
